src/scf/newton.jl

```diff
@@ -0,0 +1,317 @@
+#     retraction operators
+#       - E : builds a density matrix Eρ given a density ρ via Eρ(r,r') = δ(r,r')ρ(r),
+#       - R : builds a density ρ given a density matrix O via ρ(r) = O(r,r),
+#     then we have the relation χ0₂ = -R(Ω₄^-1)E, where χ0₂ is the two_point
+#     independent-particle susceptibility (it returns δρ from a given δV) and
+#     Ω₄ is the four-point independent-particle susceptibility.
+#
+# For further details :
+# [1] Eric Cancès, Gaspard Kemlin, Antoine Levitt. Convergence analysis of
+#     direct minimization and self-consistent iterations. SIAM Journal of Matrix
+#     Analysis and Applications, 42(1):243–274 (2021)
+#
+# We presented the algorithm in the density matrix framework, but in practice,
+# we implement it with orbitals.
+# In this framework, an element on the tangent space at ψ can be written as
+#       δP = Σ |ψi><δψi| + hc
+# where the δψi are of size Nb and are all orthogonal to the ψj, 1 <= j <= N.
+# Therefore we store them in the same kind of array than ψ, with
+# δψ[ik][:, i] = δψi for each k-point.
+# In this framework :
+#   - computing Ωδψ can be done analytically with the formula
+#         Ωδψi = H*δψi - Σj λji δψj
+#     where λij = <ψi|H|ψj>. This is one way among others to extend the
+#     definition of ΩδP = -[P_∞, [H(P_∞), δP]] to any point P on the manifold --
+#     we chose this one because it makes Ω self-adjoint for the Frobenius scalar
+#     product;
+#   - computing Kδψ can be done by constructing the δρ associated with δψ, then
+#     using the exchange-correlation kernel to compute the associated δV, then
+#     acting with δV on the ψ and projecting on the tangent space.
+
+using LinearMaps
+using IterativeSolvers
+
+#  Compute the gradient of the energy, projected on the space tangent to ψ, that
+#  is to say H(ψ)*ψ - ψ*λ where λ is the set of Rayleigh coefficients associated
+#  to the ψ.
+function compute_projected_gradient(basis::PlaneWaveBasis, ψ, occupation)
+    ρ = compute_density(basis, ψ, occupation)
+    _, H = energy_hamiltonian(basis, ψ, occupation; ρ=ρ)
+
+    [proj_tangent_kpt(H.blocks[ik] * ψ[ik], ψ[ik])
+     for ik = 1:length(basis.kpoints)]
+end
+
+# Projections on the space tangent to ψ
+function proj_tangent_kpt!(δψk, ψk)
+    # δψk = δψk - ψk * (ψk'δψk)
+    mul!(δψk, ψk, ψk'δψk, -1, 1)
+end
+proj_tangent_kpt(δψk, ψk) = proj_tangent_kpt!(copy(δψk), ψk)
+function proj_tangent(δψ, ψ)
+    [proj_tangent_kpt(δψ[ik], ψ[ik]) for ik = 1:size(ψ,1)]
+end
+function proj_tangent!(δψ, ψ)
+    [proj_tangent_kpt!(δψ[ik], ψ[ik]) for ik = 1:size(ψ,1)]
+end
+
+function apply_Ω(δψ, ψ, H::Hamiltonian, Λ)
+
+    δψ = proj_tangent(δψ, ψ)
+    Ωδψ = map(enumerate(δψ)) do (ik, δψk)
+        Hk = H.blocks[ik]
+        Λk = Λ[ik]
+
+        Ωδψk = Hk * δψk - δψk * Λk
+    end
+
+    proj_tangent!(Ωδψ, ψ)
+end
+
+# The variation in density corresponds to a variation in the orbitals.
+@views function compute_δρ(basis::PlaneWaveBasis{T}, ψ, δψ,
```

Comment:
Given the similarity of this function with compute_density, I think it makes sense to keep these two functions close in case of refactoring code. I'd vote for moving this one to `compute_density.jl` for this reason.

Reply:
Agreed. Of course what we should really do is not write this function and use AD on compute_density, but that can wait later (CC @niklasschmitz )

Reply:
Oh absolutely, good point. I'm really not yet fully in the AD mindset I realise.

Reply:
That should be no problem for ForwardDiff, once Niklas' tricks for FFTW are in. But that's a performance-critical function so it's fine having it by hand until we can do it properly (with chainrules/diffractor)